### ISSUE TITLE
url.parse misinterprets hostname

### ISSUE THREAD

**Issue by pierre-elie** [nodejs/node-v0.x-archive#8520]
Node v0.10.32

``` js
url.parse('https://good.com+.evil.org/?accessToken=xxx');
```

returns

``` js
{ 
  protocol: 'https:',
  slashes: true,
  auth: null,
  host: 'good.com',
  port: null,
  hostname: 'good.com',
  hash: null,
  search: '?accessToken=xxx',
  query: 'accessToken=xxx',
  pathname: '/+.evil.org/',
  path: '/+.evil.org/?accessToken=xxx',
  href: 'https://good.com/+.evil.org/?accessToken=xxx'
}
```

`url.parse` misinterpreted `https://good.com+.evil.org/` as `https://good.com/+.evil.org/`
If we use `url.parse` to check the validity of the hostname, the test passes, but in the browser the user is redirected to the `evil.org` website.

Other characters than `+` might do the trick too.

**Comment by mathiasbynens** [comment]
This seems fixed in Node.js v0.12.

```bash
$ nvm use 0.10
Now using node v0.10.36

$ node -e "var parsed = require('url').parse('https://good.com+.evil.com/'); console.log(parsed.host);"
good.com

$ nvm use 0.12
Now using node v0.12.0

$ node -e "var parsed = require('url').parse('https://good.com+.evil.com/'); console.log(parsed.host);"
good.com+.evil.com
```

**Comment by jasnell** [comment]
@joyent/node-coreteam ... this is fixed in v0.12, but not in v0.10. Do we want to backport?

**Comment by jasnell** [comment]
See: https://github.com/nodejs/node/issues/49
